tls/s2n_x509_validator.c

```diff
@@ -305,39 +307,41 @@ s2n_cert_validation_code s2n_x509_validator_validate_cert_chain(struct s2n_x509_
     while (s2n_stuffer_data_available(&cert_chain_in_stuffer) && sk_X509_num(validator->cert_chain_from_wire) < validator->max_chain_depth) {
         uint32_t certificate_size = 0;
 
-        S2N_ERROR_IF(s2n_stuffer_read_uint24(&cert_chain_in_stuffer, &certificate_size) < 0, S2N_ERR_CERT_UNTRUSTED);
-        S2N_ERROR_IF(certificate_size == 0 || certificate_size > s2n_stuffer_data_available(&cert_chain_in_stuffer), S2N_ERR_CERT_UNTRUSTED);
+        RESULT_GUARD_POSIX(s2n_stuffer_read_uint24(&cert_chain_in_stuffer, &certificate_size));
+        RESULT_ENSURE(certificate_size > 0, S2N_ERR_CERT_INVALID);
+        RESULT_ENSURE(certificate_size <= s2n_stuffer_data_available(&cert_chain_in_stuffer), S2N_ERR_CERT_INVALID);
 
         struct s2n_blob asn1cert = {0};
         asn1cert.size = certificate_size;
         asn1cert.data = s2n_stuffer_raw_read(&cert_chain_in_stuffer, certificate_size);
-        POSIX_ENSURE_REF(asn1cert.data);
+        RESULT_ENSURE_REF(asn1cert.data);
 
         const uint8_t *data = asn1cert.data;
 
         /* the cert is der encoded, just convert it. */
         server_cert = d2i_X509(NULL, &data, asn1cert.size);
-        S2N_ERROR_IF(!server_cert, S2N_ERR_CERT_UNTRUSTED);
+        RESULT_ENSURE_REF(server_cert);
 
         /* add the cert to the chain. */
         if (!sk_X509_push(validator->cert_chain_from_wire, server_cert)) {
             X509_free(server_cert);
```

Comment:
I don't think `server_cert` needs a cleanup function (but could be missing something). If `d2i_X509` fails, it will return null and should not allocate anything, so the `ENSURE` statement should not cause a leak. After the cert gets added to `cert_chain_from_wire`, it will be freed along with all of the other validator stuff in `s2n_x509_validator_wipe`, and is only manually freed if it doesn't get added to this list for some reason.

I added a comment about this in dbd9003d7705382cbc1364727f8e7846f83cc09d.

Reply:
Ok I just wanted to make sure that the `sk_X509_push` call took ownership. The comment helps!